ipatests/test_integration/test_uninstallation.py

```diff
@@ -89,17 +93,14 @@ def test_failed_uninstall(self):
                 '%s/%s' % (paths.ETC_DIRSRV, instance_name)
             ])
 
-            # DS has been marked as uninstalled so force the issue
-            ds.stop_tracking_certificates(serverid)
-
-            self.master.run_command([
-                paths.REMOVE_DS_PL,
-                '-i', instance_name
-            ])
-
             cmd = self.master.run_command([
                 'ipa-server-install',
                 '--uninstall', '-U'],
                 raiseonerr=False
             )
             assert cmd.returncode == 0
+
+            self.master.run_command([
+                paths.REMOVE_DS_PL,
```

Comment:
Why is it necessary to run the 389-DS uninstaller again? In case ``ipa-server-install --uninstall`` is successful, the 389-DS instance is removed. In case it fails, the assert will trigger and the uninstaller is not executed.

Reply:
It's belt-and-suspenders in case something goes wrong to not leave an ns-slapd process running.

The original purpose of the test is to ensure that on failure a non-zero error is returned. This is forced by moving the DS instance out of the way.

In the DS uninstaller (and others) use restore_state() which removes the state from the stored file so if something blows up later then it may not be possible to know that there are leftovers. While this looks bad it hasn't been a common problems in my experience. But it does mean that if remove-ds.pl fails then DS is already marked as uninstalled therefore needs cleanup manually.

Reply:
But ``remove-ds.pl`` will never get executed when ``ipa-server-install`` fails. In error case, assert fails and raises an exception. I think this should do the trick (assuming that either ipa-server-install --uninstall cleans up everything or the other two commands do).

```
            uninstall = self.master.run_command([
                'ipa-server-install',
                '--uninstall', '-U'],
                raiseonerr=False
            )
            if uninstall.returncode != 0:
                self.master.run_command([
                    paths.IPA_GETCERT,
                    'stop-tracking',
                    '-d', '%s/%s' % (paths.ETC_DIRSRV, instance_name),
                    '-n', 'Server-Cert',
                ], raiseonerr=False)

                self.master.run_command([
                    paths.REMOVE_DS_PL,
                    '-i', instance_name
                ])
                pytest.fail("ipa-server-install --uninstall has failed")

```